### PR TITLE
Merge database into Release v0.4.0

### PR DESCRIPTION
**Pull request by RainFallsSilent**

### CODE REVIEW

File: blockchain/chainstoreffldb.go [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+		return nil, err
+	}
+
+	b := new(Block)
+	err = b.Deserialize(bytes.NewReader(blkBytes))
+	if err != nil {
+		return nil, errors.New("failed to deserialize block")
+	}
+
+	if c.blocksCache != nil {
```
Comment (Houshoupei84) [inline blockchain/chainstoreffldb.go]: c.blockCache needs lock

File: blockchain/utxocache.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+			result[input] = prevTx.Outputs[input.Previous.Index]
+			up.insertReference(input, prevTx.Outputs[input.Previous.Index])
+		}
+	}
+
+	return result, nil
+}
+
+func (up *UTXOCache) GetTransaction(txID common.Uint256) (*types.Transaction, error) {
+	prevTx, exist := up.txCache[txID]
```
Comment (Houshoupei84) [inline blockchain/utxocache.go]: may needs lock

File: database/ffldb/blockio.go [hunk 3]
```diff
@@ -0,0 +1,772 @@
...
+// The function also keeps track of the open files, performs least recently
+// used tracking, and limits the number of open files to maxOpenFiles by closing
+// the least recently used file as needed.
+//
+// This function MUST be called with the overall files mutex (s.obfMutex) locked
+// for WRITES.
+func (s *blockStore) openFile(fileNum uint32) (*lockableFile, error) {
+	// Open the appropriate file as read-only.
+	filePath := blockFilePath(s.basePath, fileNum)
+	file, err := os.Open(filePath)
```
Comment (Houshoupei84) [inline database/ffldb/blockio.go]: may needs Close

File: blockchain/chainstoreffldb.go (hunk 1 above)
Comment (RainFallsSilent) [reply blockchain/chainstoreffldb.go]: fixed

File: database/ffldb/blockio.go (hunk 3 above)
Comment (RainFallsSilent) [reply database/ffldb/blockio.go]: fixed

File: blockchain/utxocache.go (hunk 2 above)
Comment (RainFallsSilent) [reply blockchain/utxocache.go]: fixed

**Pull request merged**